[Users] Last Seen Online Implementation

harriet-fisher opened the pull request:
Addresses issue #1490: Implement user online status and last seen functionality
This PR implements online status tracking with last_seen functionality for user profiles and uses middleware.py tracking to monitor user activity across the entire app.

Model & Database:
- Added last_seen DateTimeField to User model with timezone-aware tracking
- Added helper methods: is_online(), get_online_status(), and get_last_seen_display()

Middleware:
- Created LastSeenMiddleware to automatically update user's last_seen timestamp on each HTTP request

API:
- Updated FriendsTable to display online status badges and last seen information
- changed UserSerializer to include read-only fields for online_status, last_seen_display, and is_online

Frontend:
- templates/users/userprofile_detail.html displays online status ( which can either be Online/Recently Active/Offline) and timestamps
- Friends table also shows online status with color-coded badges

harriet-fisher committed:
Implemented online status tracking with last_seen field, middleware for tracking user activity, improved the chat functionality, and the migrations for the new features.

harriet-fisher committed:
Fix migration dependency for 0016_add_last_seen_field

harriet-fisher committed:
Fixed linter errors

harriet-fisher committed:
Remove unused timezone import from chat consumer after removing last_seen functionality

davdma reviewed (approved):
LGTM Harriet - all of these features are going to make the demo a more user friendly experience. Sometimes I forget that the profiles revolve around games, so having added a last online status is a really cool feature.

harriet-fisher committed:
Merge branch 'dev' into users/last-seenonline

cmarshall282 commented on src/chigame/chat/migrations/0004_remove_livechat_pinned_message.py:
```diff
@@ -0,0 +1,16 @@
+# Generated by Django 4.2.20 on 2025-05-24 00:35
```
Please remove this migration (it is not ours)

cmarshall282 commented on src/chigame/users/models.py:
```diff
@@ -73,6 +75,53 @@ def save(self, *args, **kwargs):
+        self.save(update_fields=["last_seen"])
+
+    def is_online(self, threshold_minutes=5):
+        if not self.last_seen:
```
Why do we have is_online and get_online_status that have slightly different logic? Can we avoid this?

cmarshall282 reviewed (changes requested):
Great start! Please address the minor inconsistencies in the models.

harriet-fisher commented:
To address comments:
- refactored is_online() and get_online_status() methods to have distinct purposes (+ documentation) instead of overlapping logic
- Added is_currently_online property to User model that uses the same 5-minute threshold as get_online_status so it is consistent

harriet-fisher committed:
Merge branch 'dev' into users/last-seenonline

harriet-fisher committed:
Addressed overlapping logic issues

harriet-fisher committed:
Merge branch 'users/last-seenonline' of https://github.com/uchicago-cs/chigame into users/last-seenonline

harriet-fisher committed:
Merge branch 'dev' into users/last-seenonline

harriet-fisher committed:
fixed migration errors

harriet-fisher committed:
Fixed linter

cmarshall282 reviewed (approved):
Thanks for updating that

cmarshall282 committed:
Merge branch 'dev' into users/last-seenonline

cmarshall282 commented:
Issue Score: Satisfactory

Comments:
Great job! Here is some feedback:

- Make sure that all commits follow the format specified in the ChiGame Developer Guidelines
- Please make sure that you do a thorough code review and, in the future, make sure to look out for the kind of issues the senior developer pointed out.

<!-- {"score": "satisfactory", "minor": false} -->
